js-packages/@prestojs/viewmodel/src/ViewModelCache.ts

```diff
@@ -674,12 +1095,157 @@ export default class ViewModelCache<
+                }
+            }
+            if (!sourceFieldNames.includes(sourceFieldName)) {
+                sourceFieldNames.push(sourceFieldName);
+            }
+        }
+        // If record exists under fieldNames key already then return it
+        let record = recordCache.get(fieldNames as FieldPath[]);
+        if (record) {
+            return record;
+        }
+        // Otherwise retrieve it without the related field names (ie. just look it
+        // up with the non-related field names + corresponding id's for related fields)
+        record = recordCache.get(sourceFieldNames);
+        // If record exists with the related field ids then attempt to resolve the
+        // related record
+        if (record && relations.length > 0) {
+            let cacheFieldNames: FieldPath[] = fieldNames as FieldPath[];
+            // For each request relation the corresponding record will be populated in this
+            // object. If any of the requested relations don't exist in the cache then it's
+            // considered a miss and `null` will be returned.
+            const relationData = {};
+            for (const relation of relations) {
+                // Use specified fields or if not specified default to non-related fields. This
+                // could default to '*' but it's much more difficult to handle circular references.
+                // This way if you want related records you have to be explicit.
+                const relatedFieldNames =
+                    nestedFields[relation.name] ||
+                    relation.to.fieldNames.filter(fieldName => {
+                        return !(relation.to.fields[fieldName] instanceof RelatedViewModelField);
+                    });
+                if (!nestedFields[relation.name]) {
+                    const index = cacheFieldNames.indexOf(relation.name);
+                    cacheFieldNames.splice(
+                        index,
+                        1,
+                        ...(relatedFieldNames as string[]).map(f => [relation.name, f])
+                    );
+                }
+                if (record[relation.sourceFieldName] == null) {
+                    relationData[relation.name] = null;
+                    continue;
+                }
+                let relatedRecord = relation.to.cache.get(
+                    record[relation.sourceFieldName],
+                    relatedFieldNames
+                );
+                // We had an id but the underlying record isn't available
+                if (!relatedRecord) {
+                    return null;
+                }
+                relationData[relation.name] = relatedRecord;
+            }
+            if (Object.keys(relationData).length > 0) {
+                // This could exist as cachedFieldNames can be changed above - if so we can return the
+                // record now otherwise we must create a new cache entry for the record with resolved
+                // relations.
```

Comment:
See comment above: do we need to cache this?

Reply:
so that next time you retrieve it from the cache it's the same object + don't have to go through all the logic to work out if it should exist